### PR TITLE
Transport service 

### PR DESCRIPTION
**Pull request by mloufra**
### Description
To avoid passing the object of `extensionRunner` in `ExtensionsRunner` main function, move all of the required methods and variables to `NettyTransport` file
Address comment from [previous PR](https://github.com/opensearch-project/opensearch-sdk-java/pull/166).

### Issues Resolved
https://github.com/opensearch-project/opensearch-sdk-java/pull/166
https://github.com/opensearch-project/opensearch-sdk-java/issues/116

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/sdk/NettyTransport.java [hunk 1]
```diff
@@ -40,9 +40,23 @@
...
+     * @param extensionsRunner method to call
+     */
+    public NettyTransport(ExtensionsRunner extensionsRunner) {
+        this.extensionsRunner = extensionsRunner;
+    }
+
+    /**
+     * empty constructor for getNetty4Transport method call
+     */
+    public NettyTransport() {}
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/NettyTransport.java:58]: Why is this needed?  Without initializing the `extensionsRunner` you'll end up with an NPE when you try to initialize.

File: src/main/java/org/opensearch/sdk/NettyTransport.java [hunk 2]
```diff
@@ -40,9 +40,23 @@
 
 public class NettyTransport {
     private static final String NODE_NAME_SETTING = "node.name";
+    private ExtensionsRunner extensionsRunner;
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/NettyTransport.java:43]: Suggest making this final. We want to guarantee that it's initialized in the constructor, and there will only ever be one runner set.

File: src/main/java/org/opensearch/sdk/NettyTransport.java [hunk 3]
```diff
@@ -85,22 +99,18 @@ public Netty4Transport getNetty4Transport(Settings settings, ThreadPool threadPo
...
 
         ThreadPool threadPool = new ThreadPool(settings);
 
         Netty4Transport transport = getNetty4Transport(settings, threadPool);
 
-        // Stop any existing transport service
-        if (extensionsRunner.extensionTransportService != null) {
-            extensionsRunner.extensionTransportService.stop();
-        }
+        // delete stop
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/NettyTransport.java:110]: This looks like a note-to-self to remove, so you can probably remove it.   

Note for other reviewers: I added the stop/start code myself at some point thinking it was needed, but it's not.

File: src/test/java/org/opensearch/sdk/TransportCommunicationIT.java [hunk 4]
```diff
@@ -35,7 +35,9 @@ public class TransportCommunicationIT extends OpenSearchIntegTestCase {
     private final int port = 7777;
     private final String host = "127.0.0.1";
     private volatile String clientResult;
+    private ExtensionsRunner extensionsRunner;
     private NettyTransport nettyTransport = new NettyTransport();
+    private NettyTransport nettyTransport1 = new NettyTransport(extensionsRunner);
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/sdk/TransportCommunicationIT.java:40]: You'll need to initialize the `extensionsRunner` object 2 lines up, otherwise this will fail.  And you should be deleting the no-arg constructor call so you don't need a copy with 1 appended.

**Review by dbwiddis — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensearch/sdk/NettyTransport.java (hunk 1 above)
Comment (mloufra) [reply src/main/java/org/opensearch/sdk/NettyTransport.java:58]: Before I added this empty constructor in `NettyTransport.java`, The test class `TestNetty4Transport` and `TransportCommunicationIT` shows error message "The constructor NettyTransport() is undefined".

Comment (mloufra) [reply src/main/java/org/opensearch/sdk/NettyTransport.java]: The constructor problem is fixed.  Thank @vibrantvarun Varun help me on that.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 5]
```diff
@@ -78,6 +79,7 @@ public class ExtensionsRunner {
     /**
      * This field is initialized by a call from {@link ExtensionsInitRequestHandler}.
      */
+    @Nullable
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ExtensionsRunner.java:82]: I'm on the fence over whether to use this annotation here.  This field _can_ be null until it's initialized but we don't want it to be.  So it's a good warning that we need to null-check this if we ever read it, but that should primarily be localized to this class.  Interested in other reviewers' opinions here.

I don't think it hurts to have it, I just think it's unnecessary.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java [hunk 6]
```diff
@@ -98,7 +100,7 @@ public class ExtensionsRunner {
     private ExtensionsIndicesModuleNameRequestHandler extensionsIndicesModuleNameRequestHandler =
         new ExtensionsIndicesModuleNameRequestHandler();
     private ExtensionsRestRequestHandler extensionsRestRequestHandler = new ExtensionsRestRequestHandler();
-    private NettyTransport nettyTransport = new NettyTransport();
+    private NettyTransport nettyTransport = new NettyTransport(this);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/ExtensionsRunner.java:103]: I really like this pattern for a lot of the refactorings which need to know what the ExtensionsRunner object is.

File: src/main/java/org/opensearch/sdk/NettyTransport.java [hunk 7]
```diff
@@ -40,9 +40,18 @@
 
 public class NettyTransport {
     private static final String NODE_NAME_SETTING = "node.name";
+    private final ExtensionsRunner extensionsRunner;
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/NettyTransport.java:43]: Good use of `final` here!

File: src/main/java/org/opensearch/sdk/NettyTransport.java [hunk 8]
```diff
@@ -40,9 +40,18 @@
...
     private final TransportInterceptor NOOP_TRANSPORT_INTERCEPTOR = new TransportInterceptor() {
     };
 
+    /**
+     *
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/NettyTransport.java:48]: Nit: whitespace.  

Non-nit: this javadoc isn't very explanatory.  Something like "Instantiate this object with a reference to the ExtensionsRunner."   Also the `@param` tag description doesn't describe the ExtensionsRunner.

File: src/test/java/org/opensearch/sdk/TestNetty4Transport.java [hunk 9]
```diff
@@ -25,11 +25,14 @@
...
 
     private ThreadPool threadPool;
-    private NettyTransport nettyTransport = new NettyTransport();
+    private ExtensionsRunner extensionsRunner;
+    private NettyTransport nettyTransport;
 
     @BeforeEach
     public void setUp() throws IOException {
         this.threadPool = new TestThreadPool("test");
+        this.extensionsRunner = new ExtensionsRunner();
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/sdk/TestNetty4Transport.java:34]: (Not for this PR) I wonder if there's value in creating a "Test" version of the ExtensionsRunner.  The no-arg constructor here is used exclusively for testing and I'd like to see it removed and replaced with some TestExtensionsRunner class in the test suite.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Code changes look good.  Just need some doc fixes on the constructor javadoc.

File: src/test/java/org/opensearch/sdk/TestNetty4Transport.java (hunk 9 above)
Comment (dbwiddis) [reply src/test/java/org/opensearch/sdk/TestNetty4Transport.java:34]: Created #171 for this.

File: src/main/java/org/opensearch/sdk/ExtensionsRunner.java (hunk 5 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/ExtensionsRunner.java:76]: Let's remove this @mloufra!

File: src/test/java/org/opensearch/sdk/TestNetty4Transport.java [hunk 10]
```diff
@@ -25,11 +25,13 @@
 public class TestNetty4Transport extends OpenSearchTestCase {
 
     private ThreadPool threadPool;
-    private NettyTransport nettyTransport = new NettyTransport();
+    private ExtensionsRunner extensionsRunner;
```
Comment (owaiskazi19) [inline src/test/java/org/opensearch/sdk/TestNetty4Transport.java:28]: Can we use `TestExtensionsRunner` here @dbwiddis?

Comment (dbwiddis) [reply src/test/java/org/opensearch/sdk/TestNetty4Transport.java:31]: Yes, we can, our problem here is synchronizing all the tests so they aren't all using the transport service associated with it.  I've suggested to @mloufra to use [`@ResourceLock`](https://junit.org/junit5/docs/5.3.0/api/org/junit/jupiter/api/parallel/ResourceLock.html).

Comment (mloufra) [reply src/test/java/org/opensearch/sdk/TestNetty4Transport.java:31]: > Yes, we can, our problem here is synchronizing all the tests so they aren't all using the transport service associated with it. I've suggested to @mloufra to use [`@ResourceLock`](https://junit.org/junit5/docs/5.3.0/api/org/junit/jupiter/api/parallel/ResourceLock.html).

Since the error message is from `TransportCommunicationIT`, `TestNetty4Transport` and `TestExtensionsRunner`, I added `@ResourceLock("transportService")` in each test class, but still cause same error.

**Review by dbwiddis — DISMISSED** [review]
Fixed the bind errors by moving the netty initialization out of the constructor in the Extensions Runner and backed out the "hack" solution from intervening PR.  The rest is what was ready to go before this weekend and is GTG now! :)

**Comment by ryanbogan** [comment]
That method should take a threadPool parameter too

**Comment by ryanbogan** [comment]
I can't access the branch with the changes for some reason.  The only branch I see on @mloufra 's fork is main.

**Comment by mloufra** [comment]
> I can't access the branch with the changes for some reason. The only branch I see on @mloufra 's fork is main.

I fork our sdk repo again this morning. That's why only main branch in my fork repo.

**Comment by owaiskazi19** [comment]
> I fork our sdk repo again this morning. That's why only main branch in my fork repo.
@mloufra you should close this PR and raise again with the new fork repo as you won't be able to access `transportService` branch anymore on your fork.

**Comment by mloufra** [comment]
> > I fork our sdk repo again this morning. That's why only main branch in my fork repo.
> 
> @mloufra you should close this PR and raise again with the new fork repo as you won't be able to access `transportService` branch anymore on your fork.

Thanks.
